subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1065 @@
+				K_POLL_EVENT_INITIALIZER(
+					K_POLL_TYPE_SIGNAL,
+					K_POLL_MODE_NOTIFY_ONLY,
+					&remote->read_signal),
+			};
+
+			k_sem_give(&remote->sem);
+			have_remote_sem = false;
+
+			res = k_poll(events, ARRAY_SIZE(events), K_FOREVER);
+			if (res < 0) {
+				errno = -res;
+				res = -1;
+				goto out;
+			}
+
+			res = k_sem_take(&remote->sem, K_FOREVER);
+			if (res < 0) {
+				errno = -res;
+				res = -1;
+				goto out;
+			}
+			have_remote_sem = true;
+
+			k_poll_signal_check(&remote->read_signal, &signaled,
+					    &result);
+			if (!signaled) {
+				continue;
+			}
+
+			switch (result) {
+				case SPAIR_SIG_DATA: {
+					break;
+				}
+
+				case SPAIR_SIG_CANCEL: {
+					errno = EPIPE;
+					res = -1;
+					goto out;
+				}
+
+				default: {
+					LOG_DBG("unknown signal %d received",
+						result);
+					continue;
+				}
+			}
+
+			/* SPAIR_SIG_DATA was received */
+			break;
+		}
+	}
+
+	res = k_pipe_put(&remote->recv_q, (void *)buffer, count,
+			 &bytes_written, 1, K_NO_WAIT);
+	if (res < 0) {
+		errno = -res;
+		res = -1;
+		goto out;
+	}
+
+	res = k_poll_signal_raise(&remote->write_signal, SPAIR_SIG_DATA);
+	if (res < 0) {
+		errno = -res;
+		res = -1;
+		goto out;
+	}
+
+	res = bytes_written;
+
+out:
+
+	if (have_remote_sem) {
+		k_sem_give(&remote->sem);
+	}
+	if (have_local_sem) {
+		k_sem_give(&spair->sem);
+	}
+
+	return res;
+}
+
+/**
+ * Read data from one end of a @ref spair
+ *
+ * Data written on one file descriptor of a socketpair (with e.g. write(2) or
+ * send(2)) can be read at the other end using common POSIX calls such as
+ * read(2) or recv(2).
+ *
+ * If the underlying file descriptor has the @ref O_NONBLOCK flag set then
+ * this function will return immediately. If no data was read from a
+ * non-blocking file descriptor, then -1 will be returned and @ref errno will
+ * be set to @ref EAGAIN.
+ *
+ * Blocking read operations occur when the @ref O_NONBLOCK flag is @em not set
+ * and there are no bytes to read in the @em local @ref spair.pipe.
+ *
+ * Such a blocking read will suspend execution of the current thread until
+ * one of two possible results is received on the @em local
+ * @ref spair.write_signal:
+ *
+ * 1) @ref SPAIR_SIG_DATA - data has been written to the @em local
+ *    @ref spair.pipe. Thus, allowing more data to be read.
+ *
+ * 2) @ref SPAIR_SIG_CANCEL - read of the the @em local @spair.pipe
+ *    must be cancelled for some reason (e.g. the file descriptor will be
+ *    closed imminently). In this case, the function will return -1 and set
+ *    @ref errno to @ref EINTR.
+ *
+ * @param obj the address of an @ref spair object cast to `void *`
+ * @param buffer the buffer in which to read
+ * @param count the number of bytes to read
+ *
+ * @return on success, a number > 0 representing the number of bytes written
+ * @return -1 on error, with @ref errno set appropriately.
+ */
+static ssize_t spair_read(void *obj, void *buffer, size_t count)
+{
+	ssize_t res;
```

Comment:
same comments to read function as the write one, variable declarations and empty lines after block ending }

Reply:
Would be awesome if .clang-format Just Worked (TM) .. whitespace issues should really occupy zero time of code reviewers / developers.

Reply:
Seems to be covered by uncrustify #21392

That's not a tool that I typically use in my regular workflow. It would be smart to cover it in the dev documentation and to add a pre-commit hook for it.